cpp/src/arrow/acero/asof_join_node.cc

```diff
@@ -704,21 +846,27 @@ class InputState {
         Rehash();
       }
       memo_.Store(ts, rb, latest_ref_row_, latest_time, GetLatestKey());
+      // negative tolerance means a last-known entry was stored - set `updated` to `true`
       updated = memo_.no_future_;
       ARROW_ASSIGN_OR_RAISE(advanced, Advance());
     } while (advanced);
-    if (!memo_.no_future_) {  // "updated" was not modified in the loop; set it here
+    if (!memo_.no_future_ && latest_time >= ts) {
```

Comment:
Why this change?

Reply:
This is where the exposed `latest_time` is used. If `latest_time` is less than `ts` then we know upfront that `updated = memo_.RemoveEntriesWithLesserTime(ts);` would have no effect, since all entries would have time no greater than `latest_time` and hence less than `ts`.

Reply:
Can you please add comments in the code to explain this. Looks like something could tricky to understand just from the code.